playwright/_impl/_transport.py

```diff
@@ -79,21 +121,71 @@ async def run(self) -> None:
                         buffer = buffer + data
                     else:
                         buffer = data
-                obj = json.loads(buffer)
 
-                if "DEBUGP" in os.environ:  # pragma: no cover
-                    print("\x1b[33mRECV>\x1b[0m", json.dumps(obj, indent=2))
+                obj = self.deserialize_message(buffer)
                 self.on_message(obj)
             except asyncio.IncompleteReadError:
                 break
             await asyncio.sleep(0)
         self._stopped_future.set_result(None)
 
     def send(self, message: Dict) -> None:
-        msg = json.dumps(message)
-        if "DEBUGP" in os.environ:  # pragma: no cover
-            print("\x1b[32mSEND>\x1b[0m", json.dumps(message, indent=2))
-        data = msg.encode()
+        data = self.serialize_message(message)
         self._output.write(
             len(data).to_bytes(4, byteorder="little", signed=False) + data
         )
+
+
+class WebSocketTransport(AsyncIOEventEmitter, Transport):
+    def __init__(self, ws_endpoint: str, timeout: float) -> None:
+        super().__init__()
+        Transport.__init__(self)
+
+        self._stopped = False
+        self.ws_endpoint = ws_endpoint
+        self.timeout = timeout
+        self._loop: asyncio.AbstractEventLoop
+
+    def stop(self) -> None:
+        self._stopped = True
+        self._loop.create_task(self._connection.close())
+
+    def cleanup(self) -> None:
+        self.on_error_future.cancel()
+
+    async def wait_until_stopped(self) -> None:
+        await self._connection.wait_closed()
+
+    async def run(self) -> None:
+        await super().run()
+
+        options = {}
+        if self.timeout is not None:
+            options["close_timeout"] = self.timeout / 1000
+            options["ping_timeout"] = self.timeout / 1000
+        self._connection = await websockets.connect(self.ws_endpoint, **options)
+
+        while not self._stopped:
+            try:
+                message = await self._connection.recv()
+                obj = self.deserialize_message(message)
+                self.on_message(obj)
+            except websockets.exceptions.ConnectionClosed:
+                if not self._stopped:
+                    self.emit("close")
+                self.on_error_future.set_exception(
+                    Error("Playwright connection closed")
+                )
+                break
+            except Exception as exc:
+                print(f"Received unhandled exception: {exc}")
+                self.on_error_future.set_exception(exc)
+
+    def send(self, message: Dict) -> None:
+        self.test_if_closed()
+        data = self.serialize_message(message)
+        self._loop.create_task(self._connection.send(data))
+
+    def test_if_closed(self) -> None:
+        if self._stopped or self._connection.closed:
```

Comment:
You can skip this one: It would be better to inline this method as it is used only once